File: src/android/MapboxSecurity.java
package com.outsystems.mapbox;

import java.net.URI;

/** Input policy shared by every native entry point. No Android dependencies. */
final class MapboxSecurity {
    static final int MAX_POINTS = 20000;
    static final int MAX_INPUT_CHARS = 4 * 1024 * 1024;
    static final double MAX_TILES = 50000;

    static boolean publicToken(String token) {
        return token != null && token.matches("pk\\.[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+");
    }

    static boolean styleAllowed(String value, String hosts) {
        if (value == null || value.length() > 4096) return false;
        if (value.matches("mapbox://styles/[A-Za-z0-9_-]+/[A-Za-z0-9_-]+")) return true;
        try {
            URI uri = new URI(value);
            if (!"https".equalsIgnoreCase(uri.getScheme()) || uri.getHost() == null
                    || uri.getRawUserInfo() != null || uri.getRawFragment() != null
                    || (uri.getPort() != -1 && uri.getPort() != 443)) return false;
            for (String host : hosts.split(",")) {
                if (uri.getHost().equalsIgnoreCase(host.trim())) return true;
            }
        } catch (Exception ignored) { }
        return false;
    }

    static boolean validZoom(double zoom) {
        return Double.isFinite(zoom) && zoom >= 2 && zoom <= 18;
    }

    // Conservative bounding-box budget, including lower zoom levels. Reject
    // antimeridian-crossing boxes instead of underestimating their download cost.
    static boolean offlineBounds(double south, double west, double north, double east, double zoom) {
        if (!validZoom(zoom) || !Double.isFinite(south) || !Double.isFinite(west)
                || !Double.isFinite(north) || !Double.isFinite(east)
                || south < -85 || north > 85 || west < -180 || east > 180
                || south > north || west > east || east - west > 180) return false;
        if (distance(south, west, north, east) > 100000) return false;
        double scale = Math.pow(2, Math.ceil(zoom));
        double columns = Math.ceil((east - west) / 360 * scale) + 2;
        double rows = Math.ceil(Math.abs(mercator(north) - mercator(south)) * scale) + 2;
        return columns * rows * 4 / 3 <= MAX_TILES;
    }

    private static double mercator(double lat) {
        double rad = Math.toRadians(lat);
        return (1 - Math.log(Math.tan(rad) + 1 / Math.cos(rad)) / Math.PI) / 2;
    }

    private static double distance(double lat1, double lon1, double lat2, double lon2) {
        double a = Math.pow(Math.sin(Math.toRadians(lat2 - lat1) / 2), 2)
            + Math.cos(Math.toRadians(lat1)) * Math.cos(Math.toRadians(lat2))
            * Math.pow(Math.sin(Math.toRadians(lon2 - lon1) / 2), 2);
        return 6371000 * 2 * Math.asin(Math.sqrt(Math.min(1, a)));
    }
}

File: src/ios/MapboxSecurity.swift
import Foundation

enum MapboxSecurity {
    static let maxPoints = 20000
    static let maxInputBytes = 4 * 1024 * 1024

    static func publicToken(_ token: String) -> Bool {
        token.range(of: "^pk\\.[A-Za-z0-9_-]+\\.[A-Za-z0-9_-]+$", options: .regularExpression) != nil
    }

    static func styleAllowed(_ value: String, hosts: String) -> Bool {
        guard value.utf8.count <= 4096 else { return false }
        if value.range(of: "^mapbox://styles/[A-Za-z0-9_-]+/[A-Za-z0-9_-]+$", options: .regularExpression) != nil { return true }
        guard let url = URLComponents(string: value), url.scheme?.lowercased() == "https",
              let host = url.host, url.user == nil, url.password == nil,
              url.fragment == nil, url.port == nil || url.port == 443 else { return false }
        return hosts.split(separator: ",").contains {
            $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == host.lowercased()
        }
    }

    static func zoom(_ value: Double) -> UInt8? {
        guard value.isFinite, value >= 2, value <= 18 else { return nil }
        return UInt8(value)
    }

    static func offlineBounds(south: Double, west: Double, north: Double, east: Double, zoom: Double) -> Bool {
        guard self.zoom(zoom) != nil, [south, west, north, east].allSatisfy({ $0.isFinite }),
              south >= -85, north <= 85, west >= -180, east <= 180,
              south <= north, west <= east, east - west <= 180 else { return false }
        let rad = Double.pi / 180
        let a = pow(sin((north - south) * rad / 2), 2)
            + cos(south * rad) * cos(north * rad) * pow(sin((east - west) * rad / 2), 2)
        guard 6371000 * 2 * asin(sqrt(min(1, a))) <= 100000 else { return false }
        func mercator(_ lat: Double) -> Double {
            (1 - log(tan(lat * rad) + 1 / cos(lat * rad)) / Double.pi) / 2
        }
        let scale = pow(2, ceil(zoom))
        let columns = ceil((east - west) / 360 * scale) + 2
        let rows = ceil(abs(mercator(north) - mercator(south)) * scale) + 2
        return columns * rows * 4 / 3 <= 50000
    }
}
